File: euler15.py
# Сколько существует таких маршрутов в сетке 20×20?

from datetime import datetime
from functools import lru_cache
from math import factorial
from typing import Dict, Tuple


memory: Dict[Tuple[int, int], int] = {(1, 0): 1, (0, 1): 1}


@lru_cache(2 ** 20)
def func(x: int, y: int) -> int:
    """
    Возвращает количество маршрутов до точки (x, y).

    Args:
        x: Координата x.
        y: Координата y.

    Returns:
        Количество маршрутов.
    """
    if (x, y) not in memory:
        if x == 0:
            memory[(x, y)] = func(x, y - 1)
        elif y == 0:
            memory[(x, y)] = func(x - 1, y)
        else:
            memory[(x, y)] = func(x - 1, y) + func(x, y - 1)
    return memory[(x, y)]


def euler15() -> None:
    """
    Решение задачи Эйлера №15.

    Сколько существует таких маршрутов в сетке 20×20?
    """
    n = 20

    # 1 вариант
    start_time = datetime.now()
    print(func(n, n))
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    counter = 1
    for i in range(n):
        counter = counter * (2 * n - i) // (1 + i)
    print(counter)
    print(datetime.now() - start_time)

    # 3 вариант
    start_time = datetime.now()
    print(int(factorial(2 * n) / (factorial(n) ** 2)))
    print(datetime.now() - start_time)

    # 4 вариант
    start_time = datetime.now()
    dp = {(0, 0): 1}
    for i in range(n + 1):
        for j in range(n + 1):
            if i == 0 or j == 0:
                dp[(i, j)] = 1
            else:
                dp[(i, j)] = dp[(i - 1, j)] + dp[(i, j - 1)]
    print(dp[(n, n)])
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler15()

File: go/euler15.go
// euler15 находит количество маршрутов в сетке 20×20.
//
// Решение задачи Эйлера №15.
package main

import (
	"fmt"
	"time"
)

var memory = make(map[[2]int]int)

func f(x, y int) int {
	key := [2]int{x, y}
	if v, ok := memory[key]; ok {
		return v
	}
	var result int
	if x == 0 {
		result = f(x, y-1)
	} else if y == 0 {
		result = f(x-1, y)
	} else {
		result = f(x-1, y) + f(x, y-1)
	}
	memory[key] = result
	return result
}

func euler15() {
	n := 20

	memory = make(map[[2]int]int)
	memory[[2]int{1, 0}] = 1
	memory[[2]int{0, 1}] = 1

	startTime := time.Now()
	fmt.Println(f(n, n))
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	counter := 1
	for i := 0; i < n; i++ {
		counter = counter * (2*n - i) / (1 + i)
	}
	fmt.Println(counter)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	dp := make(map[[2]int]int)
	for i := 0; i <= n; i++ {
		for j := 0; j <= n; j++ {
			if i == 0 || j == 0 {
				dp[[2]int{i, j}] = 1
			} else {
				dp[[2]int{i, j}] = dp[[2]int{i - 1, j}] + dp[[2]int{i, j - 1}]
			}
		}
	}
	fmt.Println(dp[[2]int{n, n}])
	fmt.Println(time.Since(startTime))
}

func main() {
	euler15()
}
